### ISSUE TITLE
Weird interaction of jit and functools.partial

### ISSUE THREAD

**Issue by proteneer** [google/jax#731]
``` python
import unittest
import functools

import numpy as onp
import jax.numpy as jnp
import jax

def debug_fn(conf, params):
    energy = jnp.sum(jnp.power(conf - params, 2.0)) 
    return energy

def loss_fn(p, c):
    energy_fn = functools.partial(debug_fn, params=p)

    # grad_fn = jax.grad(energy_fn, argnums=(0,)) # works
    grad_fn = jax.jit(jax.grad(energy_fn, argnums=(0,))) # broken: can't lift sublevels    

    return grad_fn(c)[0]

class TestForcefield(unittest.TestCase):

    def test_minimization(self):
        loss_grad_fn = jax.jacfwd(loss_fn, argnums=(0,))
        loss_grad = loss_grad_fn(0.5, 1.0)[0]
```
Turning on jit for a wrapped partial returns an incompatible sub-level error:

```
ff/jax_debug2.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../venv/lib/python3.6/site-packages/jax/api.py:332: in jacfun
    y, jac = vmap(pushfwd, out_axes=(None, -1))(_std_basis(dyn_args))
../../venv/lib/python3.6/site-packages/jax/api.py:481: in batched_fun
    out_flat = batching.batch(jaxtree_fun, in_flat, in_axes_, out_axes)
../../venv/lib/python3.6/site-packages/jax/interpreters/batching.py:45: in batch
    return batch_transform(fun, sz, in_dims, out_dim_dst).call_wrapped(in_vals)
../../venv/lib/python3.6/site-packages/jax/linear_util.py:151: in call_wrapped
    ans = gen.send(ans)
../../venv/lib/python3.6/site-packages/jax/api_util.py:46: in pytree_fun_to_jaxtupletree_fun
    yield pytree_to_jaxtupletree(ans)
../../venv/lib/python3.6/site-packages/jax/tree_util.py:112: in process_pytree
    return walk_pytree(process_node, lambda x: x, tree)
../../venv/lib/python3.6/site-packages/jax/tree_util.py:122: in walk_pytree
    return f_node(proc_children), tree_def
../../venv/lib/python3.6/site-packages/jax/core.py:583: in pack
    return pack_p.bind(*args)
../../venv/lib/python3.6/site-packages/jax/core.py:603: in pack_p_bind
    tracers = map(top_trace.full_raise, args)
../../venv/lib/python3.6/site-packages/jax/util.py:43: in safe_map
    return list(map(f, *args))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = BatchTrace(level=0/0), val = Traced<ShapedArray(float32[]):JaxprTrace(level=-1/1)>

    def full_raise(self, val):
      if not isinstance(val, Tracer):
        return self.pure(val)
      level = self.level
      sublevel = self.sublevel
      if val.trace.master is self.master:
        if val.trace.sublevel == sublevel:
          return val
        elif val.trace.sublevel < sublevel:
          return self.sublift(val)
        else:
          raise Exception("Can't lift sublevels {} to {}"
                          .format(val.trace.sublevel, sublevel))
      elif val.trace.level < level:
        if val.trace.sublevel > sublevel:
          raise Exception("Incompatible sublevel: {}, {}"
>                         .format(val.trace, (level, sublevel)))
E         Exception: Incompatible sublevel: JaxprTrace(level=-1/1), (0, 0)

../../venv/lib/python3.6/site-packages/jax/core.py:231: Exception
```
This only happens with JIT turned on. I guess I'm not supposed to be mixing the two?

**Comment by dougalm** [comment]
Looks like it's just a typo in `BatchTrace.post_process_call`. Should be fixed by 7c34858. We ought to add a test too...

**Comment by mattjj** [comment]
You know a bug is scary when @dougalm takes it. This one ended up being too easy!
